ls1intum/Artemis
# `Development`: Cache active users in metrics calculation to improve performance

### Checklist
#### General


#### Server

### Motivation and Context
Certain metrics exposed to other systems (such as Grafana or Kubernetes) rely on active users (users with a submission within the last 14 days) to e.g. predict the number of students that will attend an upcoming exercise.
This calculation is relativly complex (takes around 2.5s on production) but generates results that do not change rapidly.
To further improve the performance of Artemis and to decrase database load, we should simplify the query and cache it.

### Description
This PR
- simplifies the calculation by only extraction active users (including users that submitted in test courses), not complete statistics including submission dates
- caches the list of active users for 60 minutes.

The active user list is calculated initially during the startup of Artemis (only on instances with the profile `scheduling`!) and then every 60 minutes.
The metrics are calculated every 5 minutes.

### Steps for Testing

Prerequisites:
- Artemis instance with `scheduling` profile
- Access to the metrics at `ARTEMIS-URL/management/prometheus` (either by running Artemis locally or by configuring the configuration value `spring.prometheus.monitoringIp` to a value including your IP address
- A course with (at least) 2 users: one of them should have made a submission in Artemis within the last 14 days, the other one should not

1. Create a new quiz exercise in the course that starts in 10 minutes
2. Wait 5 minutes
3. Navigate to `ARTEMIS-URL/management/prometheus` 
4. Check the property `artemis.scheduled.exercises.due.student_multiplier.active.14` (with `exerciseType`=`QUIZ`) and remember the value
5. Create a submission (in any course/exercise) with the student that previously did not have a submission within the last hour
6. Wait 60 minutes
7. Create a new quiz exercise in the course that starts in 10 minutes
8. Wait 5 minutes
9. Navigate to `ARTEMIS-URL/management/prometheus` 
10. Check the property `artemis.scheduled.exercises.due.student_multiplier.active.14` (with `exerciseType`=`QUIZ`) and ensure that the value is 1 higher then the value from step 4.

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
Note: the class `MetricsBean` contains several lines that can not easily be tested within unit tests because it relies on WebSocket statistics which are not available during tests.

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| MetricsBean.java | 86% | ✅                           |
